### PR TITLE
Oes 34

### PR DESCRIPTION
**Pull request by sneha-nagole**
**Summary:**
Refactored & slightly modified user registration process. Simplified validations to suit new registration  flow.

**What is new:**
New endpoint to verify email address. Returns 200 if email is valid & not already in use in the system & 400 if email is invalid or already in use.

**What is changed:**
Simplified validations, simplified exception handling, testable code. Spring beans injection into UserRegistrationService.

In the new registration flow:
Instead of multiple contacts, just adding email & phone number (office phone)
Abstracted notifications into its own classes

### CODE REVIEW

**Comment by tparkergh** [comment]
Looks good. Great refactoring and tests additions. Just pointing out a few minor things.

api-rest-common/src/main/java/edu/mit/ll/em/api/rs/RegisterUserNew.java
Is "New" a reference to a new object or is it part of the domain? Unless it's part of a domain concept, I would probably drop it.

One of the Junit tests had a lot of test assertions in a single method. I would probably name them a bit different and break them up. This would allow us to better identify the problem by which test broke. Other tests looked great.

api-rest-service/src/main/java/edu/mit/ll/em/api/service/RegistrationService.java
Had some commented out code. Can we delete it?

api-rest-service/src/main/java/edu/mit/ll/em/api/notification/Notification.java
This class and it's factory was empty. Was this intentional? 

The above is just minor nitpicking but overall, looks very good. I'm accepting.

**Comment by sneha-nagole** [comment]
@tparkergh : 
Thank you for the feedback.
1. Forgot to delete api-rest-common/src/main/java/edu/mit/ll/em/api/rs/RegisterUserNew.java. Will delete this.
2. I will refactor Junit tests one more time & make changes
3. api-rest-service/src/main/java/edu/mit/ll/em/api/service/RegistrationService.java: This file is renamed to UserRegistrationService.java. May be you are looking at old code. Can you please pull from remote repo one more time & let me know if there is anything that needs to be changed.
4. api-rest-service/src/main/java/edu/mit/ll/em/api/notification/Notification.java: Deleted this file as i was not using it.

File: api-rest-common/src/main/java/edu/mit/ll/em/api/rs/RegisterUser.java [hunk 1]
```diff
@@ -30,161 +30,117 @@
...
+    @NotBlank(message="Please provide Last Name.")
+    private String lastName;
+    @NotEmpty(message="Please provide valid Email Address.")
+    @Email(message="Please provide valid Email Address.")
+    private String email;
+    @Pattern(regexp="^(\\(\\d{3}\\) \\d{3}-\\d{4})?$", message="Please provide valid Phone number.")
+    private String phone;
+    @NotNull(message="Please provide Password of 8-20 characters with at least one digit, one upper case letter, one lower case letter and one special symbol @#$%-_!.")
+    @Pattern(regexp="^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@#$%_!-])[a-zA-Z0-9@#$%_!-]{8," +
+            "20}$",
```
Comment (comeara) [inline api-rest-common/src/main/java/edu/mit/ll/em/api/rs/RegisterUser.java]: Super minor but why the concatenation of strings in the regexp attribute?

File: api-rest-common/src/main/java/edu/mit/ll/em/api/rs/RegisterUserNew.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    @Pattern(regexp="^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@#$%_!-])[a-zA-Z0-9@#$%_!-]{8," +
+            "20}$",
+            message="Please provide Password of 8-20 characters with at least one digit, one upper case letter, one lower case letter and one special symbol @#$%-_!.")
+    private String password;
+    private Set<Integer> teams = new HashSet<Integer>();
+
+    public RegisterUserNew() {
+    }
+
+    public RegisterUserNew(Integer organizationTypeId, Integer organizationId, String firstName, String lastName, String email, String phone, String password, Collection<Integer> teams) {
```
Comment (comeara) [inline api-rest-common/src/main/java/edu/mit/ll/em/api/rs/RegisterUserNew.java]: I believe @tparkergh mentioned, but, this appears to be a duplicate of RegisterUser.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/configuration/SpringConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import edu.mit.ll.nics.nicsdao.impl.OrgDAOImpl;
+import edu.mit.ll.nics.nicsdao.impl.UserDAOImpl;
+import edu.mit.ll.nics.nicsdao.impl.UserOrgDAOImpl;
+import edu.mit.ll.nics.nicsdao.impl.WorkspaceDAOImpl;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import javax.validation.Validation;
+import javax.validation.Validator;
+import javax.validation.ValidatorFactory;
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/configuration/SpringConfiguration.java]: This import is not used.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+package edu.mit.ll.em.api.notification;
+
+import edu.mit.ll.em.api.util.APIConfig;
+import edu.mit.ll.em.api.util.APILogger;
+import edu.mit.ll.nics.common.email.JsonEmail;
+import edu.mit.ll.nics.common.entity.Org;
+import edu.mit.ll.nics.common.entity.User;
+import edu.mit.ll.nics.common.rabbitmq.RabbitFactory;
+import edu.mit.ll.nics.common.rabbitmq.RabbitPubSubProducer;
+import edu.mit.ll.nics.nicsdao.impl.OrgDAOImpl;
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java]: This import is not used.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.commons.configuration.Configuration;
+import org.springframework.util.StringUtils;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java]: This import is not used.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        this.emApiConfiguration = emApiConfiguration;
+    }
+
+    public NotifyFailedUserRegistration(Configuration emApiConfiguration, RabbitPubSubProducer rabbitProducer) {
+        this.emApiConfiguration = emApiConfiguration;
+        this.rabbitProducer = rabbitProducer;
+    }
+
+    public void notify(User user, Org org) {
+        String newRegisteredUsers = emApiConfiguration.getString(APIConfig.NEW_REGISTERED_USER_EMAIL);
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java]: This variable is not used.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        builder.append("\n\nA new user has attempted to register. However, their system user failed to successfully persist,");
+        builder.append(" so before they can try to register again with the same email address, their");
+        builder.append(" Identity user will need deleted in OpenAM.");
+        builder.append("\n\nName: " + user.getFirstname() + " " + user.getLastname());
+        builder.append("\nOrganization: " + org.getName());
+        builder.append("\nEmail: " + user.getUsername());
+        return builder.toString();
+    }
+
+    public String getFromEmail() {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java]: Could be private.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        return this.emApiConfiguration.getString(APIConfig.NEW_USER_ALERT_EMAIL);
+    }
+
+    public String getAlertTopic() {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java]: Could be private.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        return this.emApiConfiguration.getString(APIConfig.EMAIL_ALERT_TOPIC, "iweb.nics.email.alert");
+    }
+
+    public String getSysAdminsEmail() {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistration.java]: Could be private.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifySuccessfulUserRegistration.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        String date = new SimpleDateFormat("EEE MMM d HH:mm:ss z yyyy").format(new Date());
+        builder.append(date);
+        builder.append("\n\nA new user has registered: " + user.getUsername());
+        builder.append("\nName: " + user.getFirstname() + " " + user.getLastname());
+        builder.append("\nOrganization: " + org.getName());
+        builder.append("\nEmail: " + user.getUsername());
+        return builder.toString();
+    }
+
+    public String getFromEmail() {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifySuccessfulUserRegistration.java]: Could be private.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifySuccessfulUserRegistration.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        return this.emApiConfiguration.getString(APIConfig.NEW_USER_ALERT_EMAIL);
+    }
+
+    public String getAlertTopic() {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/notification/NotifySuccessfulUserRegistration.java]: Could be private.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/rs/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -110,17 +86,22 @@
 	private static String SUCCESS = "success";
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/rs/impl/UserServiceImpl.java]: I do not think the constants on lines 76-87 are used any longer. Can they be removed?

File: api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+        this.orgDao = orgDao;
+        this.userOrgDao = userOrgDao;
+        this.workspaceDao = workspaceDao;
+        this.openAmGatewayFactory = openAmGatewayFactory;
+        this.rabbitProducer = rabbitProducer;
+        this.successfulUserRegistrationNotification = successfulUserRegistrationNotification;
+        this.failedUserRegistrationNotification = failedUserRegistrationNotification;
+    }
+
+    private RabbitPubSubProducer getRabbitProducer() throws IOException {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java]: I cannot find where this method is called. Am I missing it?

File: api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+                rabbitProducer = RabbitFactory.makeRabbitPubSubProducer(
+                APIConfig.getInstance().getConfiguration().getString(APIConfig.RABBIT_HOSTNAME_KEY),
+                APIConfig.getInstance().getConfiguration().getString(APIConfig.RABBIT_EXCHANGENAME_KEY),
+                APIConfig.getInstance().getConfiguration().getString(APIConfig.RABBIT_USERNAME_KEY),
+                APIConfig.getInstance().getConfiguration().getString(APIConfig.RABBIT_USERPWD_KEY));
+        }
+        return rabbitProducer;
+    }
+
+    public Response postUser(int workspaceId, RegisterUser registerUser) {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java]: I don't believe the workspaceId parameter is used.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+        if(CollectionUtils.isEmpty(workspaceIds))
+            throw new Exception("No workspaces found during user registration.");
+        for(UserOrg userOrgTeam : userOrgTeams) {
+            userOrgWorkspacesTeams.addAll(createUserOrgWorkspaceEntities(userOrgTeam, false, workspaceIds));
+        }
+        return userOrgWorkspacesTeams;
+    }
+
+
+    public UserOrg createUserOrg(int orgId, int userId) throws Exception {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java]: Access can be private.

File: api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+            userorg.setCreated(Calendar.getInstance().getTime());
+            userorg.setSystemroleid(SADisplayConstants.USER_ROLE_ID);
+            userorg.setUserid(userId);
+        } else {
+            logger.e(CNAME, "Unable to get valid userOrgId from DB");
+        }
+        return userorg;
+    }
+
+    private Collection<UserOrgWorkspace> createUserOrgWorkspaceEntities(UserOrg userOrg, boolean enabled, List<Integer> workspaceIds) {
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java]: The enabled parameter is always false. Can it be removed?

File: api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java [hunk 17]
```diff
@@ -0,0 +1,343 @@
...
+            user.setUsername(rUser.getEmail());
+            user.setUserId(userId);
+            user.setPasswordHash(SaltedHashUtil.generateSaltedHash(rUser.getPassword(), rUser.getEmail()));
+            user.setEnabled(false);
+            user.setActive(true);
+            user.setLastupdated(currentTime);
+            user.setCreated(currentTime);
+            user.setPasswordchanged(currentTime);
+        } else {
+            throw new RuntimeException("Failed User registraiton. Unable to fetch a new UserId from DB.");
```
Comment (comeara) [inline api-rest-service/src/main/java/edu/mit/ll/em/api/service/UserRegistrationService.java]: Typo in the word registration.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeUtils;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java]: Unused import.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.never;
+import static org.mockito.Matchers.any;
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java]: Unused import.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+import static org.mockito.Matchers.eq;
+
+public class NotifyFailedUserRegistrationTest {
+    private NotifyFailedUserRegistration notifyFailedUserRegistration;
+    //private OrgDAOImpl orgDao = mock(OrgDAOImpl.class);
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java]: Please remove commented out code.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    private Configuration emApiConfiguration = mock(Configuration.class);
+    private RabbitPubSubProducer rabbitProducer = mock(RabbitPubSubProducer.class);
+    private User user = mock(User.class);
+    private Org org = mock(Org.class);
+    private static final int orgId = 1;
+    private static final String fromEmailAddress = "from@happytest.com";
+    private static final String adminEmailAddress = "testadmin@happy.com";
+    //private static List<String> orgAdminList = Arrays.asList(adminEmailAddress);
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifyFailedUserRegistrationTest.java]: Please remove commented out code.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifySuccessfulUserRegistrationTest.java [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    private static final String alertTopic = "alertTopic";
+    private static String hostname;
+    private static final String userFirstName = "first";
+    private static final String userLastName = "last";
+    private static final String userEmailAddress = "useremail@happy.com";
+    private static final String orgName = "orgName";
+    private NotifySuccessfulUserRegistration notifySuccessfulUserRegistration;
+
+    @Before
+    public void setup() throws UnknownHostException, IOException {
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/notification/NotifySuccessfulUserRegistrationTest.java]: Doesn't actually throw UnknownHostException. Consider declaring throws Exception instead.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/rs/exceptionMapper/ValidationExceptionMapperTest.java [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package edu.mit.ll.em.api.rs.exceptionMapper;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/rs/exceptionMapper/ValidationExceptionMapperTest.java]: Unused import.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/rs/exceptionMapper/ValidationExceptionMapperTest.java [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+import edu.mit.ll.em.api.rs.RegisterUserNew;
+import edu.mit.ll.em.api.rs.ValidationErrorResponse;
+import edu.mit.ll.em.api.rs.exceptionMapper.ValidationExceptionMapper;
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/rs/exceptionMapper/ValidationExceptionMapperTest.java]: Unused import.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/rs/exceptionMapper/ValidationExceptionMapperTest.java [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import static org.junit.Assert.*;
+
+import javax.validation.*;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/rs/exceptionMapper/ValidationExceptionMapperTest.java]: Unused import.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/rs/impl/UserServiceImplTest.java [hunk 26]
```diff
@@ -0,0 +1,7 @@
+package edu.mit.ll.em.api.rs.impl;
+
+/**
+ * Created by snagole on 3/23/17.
+ */
+public class UserServiceImplTest {
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/rs/impl/UserServiceImplTest.java]: Can this empty test be removed?

File: api-rest-service/src/test/java/edu/mit/ll/em/api/service/UserRegistrationServiceTest.java [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+        when(workspaceDao.getWorkspaceIds()).thenReturn(workspaceIds);
+        when(userDao.getContactTypeId(SADisplayConstants.EMAIL_TYPE)).thenReturn(emailContactTypeId);
+        when(userDao.getContactTypeId(SADisplayConstants.PHONE_OFFICE_TYPE)).thenReturn(officePhoneContactTypeId);
+        when( openAmGateway.createIdentityUser(any(User.class), eq(registerUser)) ).thenReturn(successJson);
+        when(openAmGateway.deleteIdentityUser(registerUser.getEmail())).thenReturn(successJson);
+        when(userDao.registerUser(any(User.class), any(List.class), any(List.class), any(List.class))).thenReturn(false);
+        Response response = service.postUser(workspaceId, registerUser);
+
+        verify(logger).i(CNAME, String.format("User registration failed for user with email address %s. Successfully rolled back changes from OpenAm", registerUser.getEmail()));
+        //verify(successfulUserRegistrationNotification).notify(any(User.class), any(Org.class));
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/service/UserRegistrationServiceTest.java]: Please remove commented code.

File: api-rest-service/src/test/java/edu/mit/ll/em/api/service/UserRegistrationServiceTest.java [hunk 28]
```diff
@@ -0,0 +1,351 @@
...
+        when(userDao.getContactTypeId(SADisplayConstants.PHONE_OFFICE_TYPE)).thenReturn(officePhoneContactTypeId);
+        when( openAmGateway.createIdentityUser(any(User.class), eq(registerUser)) ).thenReturn(successJson);
+        when(openAmGateway.deleteIdentityUser(registerUser.getEmail())).thenReturn(failureJson);
+        when(userDao.registerUser(any(User.class), any(List.class), any(List.class), any(List.class))).thenReturn(false);
+
+        Response response = service.postUser(workspaceId, registerUser);
+
+        verify(failedUserRegistrationNotification).notify(any(User.class), any(Org.class));
+        verify(logger).e(CNAME, String.format("User Registration failure for user %s, Unable to rollback changes from OpenAm", registerUser.getEmail()));
+        //verify(successfulUserRegistrationNotification).notify(any(User.class), any(Org.class));
```
Comment (comeara) [inline api-rest-service/src/test/java/edu/mit/ll/em/api/service/UserRegistrationServiceTest.java]: Please remove commented code.

**Pull request merged**